clang-tools-extra/clang-tidy/readability/UseStdMinMaxCheck.cpp

```diff
@@ -0,0 +1,132 @@
+    : ClangTidyCheck(Name, Context),
+      IncludeInserter(Options.getLocalOrGlobal("IncludeStyle",
+                                               utils::IncludeSorter::IS_LLVM),
+                      areDiagsSelfContained()),
+      AlgorithmHeader(Options.get("AlgorithmHeader", "<algorithm>")) {}
+
+void UseStdMinMaxCheck::storeOptions(ClangTidyOptions::OptionMap &Opts) {
+  Options.store(Opts, "IncludeStyle", IncludeInserter.getStyle());
+}
+
+void UseStdMinMaxCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      ifStmt(
+          has(binaryOperator(
+              anyOf(hasOperatorName("<"), hasOperatorName(">"),
+                    hasOperatorName("<="), hasOperatorName(">=")),
+              hasLHS(expr().bind("lhsVar1")), hasRHS(expr().bind("rhsVar1")))),
+          hasThen(
+              anyOf(stmt(binaryOperator(hasOperatorName("="),
+                                        hasLHS(expr().bind("lhsVar2")),
+                                        hasRHS(expr().bind("rhsVar2")))),
+                    compoundStmt(has(binaryOperator(
+                        hasOperatorName("="), hasLHS(expr().bind("lhsVar2")),
+                        hasRHS(expr().bind("rhsVar2"))))))))
+          .bind("ifStmt"),
+      this);
+}
+
+void UseStdMinMaxCheck::registerPPCallbacks(const SourceManager &SM,
+                                            Preprocessor *PP,
+                                            Preprocessor *ModuleExpanderPP) {
+  IncludeInserter.registerPreprocessor(PP);
+}
+
+void UseStdMinMaxCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto *lhsVar1 = Result.Nodes.getNodeAs<Expr>("lhsVar1");
+  const auto *rhsVar1 = Result.Nodes.getNodeAs<Expr>("rhsVar1");
+  const auto *lhsVar2 = Result.Nodes.getNodeAs<Expr>("lhsVar2");
+  const auto *rhsVar2 = Result.Nodes.getNodeAs<Expr>("rhsVar2");
+  const auto *ifStmt = Result.Nodes.getNodeAs<IfStmt>("ifStmt");
+  auto &Context = *Result.Context;
+  const SourceManager &Source = Context.getSourceManager();
+
+  if (!lhsVar1 || !rhsVar1 || !lhsVar2 || !rhsVar2 || !ifStmt)
+    return;
+
+  const auto *binaryOp = dyn_cast<BinaryOperator>(ifStmt->getCond());
+  if (!binaryOp)
+    return;
+
+  SourceLocation ifLocation = ifStmt->getIfLoc();
+  SourceLocation thenLocation = ifStmt->getEndLoc();
+
+  auto lhsVar1Str =
+      Lexer::getSourceText(Source.getExpansionRange(lhsVar1->getSourceRange()),
+                           Context.getSourceManager(), Context.getLangOpts());
+
+  auto lhsVar2Str =
+      Lexer::getSourceText(Source.getExpansionRange(lhsVar2->getSourceRange()),
+                           Context.getSourceManager(), Context.getLangOpts());
+
+  auto rhsVar1Str =
+      Lexer::getSourceText(Source.getExpansionRange(rhsVar1->getSourceRange()),
+                           Context.getSourceManager(), Context.getLangOpts());
+
+  auto replacementMax = lhsVar2Str.str() + " = std::max(" + lhsVar1Str.str() +
```

Comment:
no need to read text & produce strings if we do not enter into any if bellow.
you could wrap this into two local lambdas, and call it/create string only when needed.